### ISSUE TITLE
DOC: Explain how to continue the conversation with OllamaFunctions

### ISSUE THREAD

**Issue by tonylampada** [langchain-ai/langchain#14360]
### Issue with current documentation:

https://python.langchain.com/docs/integrations/chat/ollama_functions

Ollama functions page shows how to start a conversation where the llm calls a function.
But it stops there. 

How do we keep the conversation going? (i.e. I should give the llm the answer to the function call, and then it should give a text reply with his explanation of the weather in Paris, right?)

Here's what I tried

```
>>> from langchain_experimental.llms.ollama_functions import OllamaFunctions
>>> from langchain.schema import HumanMessage, FunctionMessage
>>> import json
>>> model = OllamaFunctions(model="mistral")
>>> model = model.bind(
...     functions=[
...         {
...             "name": "get_current_weather",
...             "description": "Get the current weather in a given location",
...             "parameters": {
...                 "type": "object",
...                 "properties": {
...                     "location": {
...                         "type": "string",
...                         "description": "The city and state, " "e.g. San Francisco, CA",
...                     },
...                     "unit": {
...                         "type": "string",
...                         "enum": ["celsius", "fahrenheit"],
...                     },
...                 },
...                 "required": ["location"],
...             },
...         }
...     ],
...     function_call={"name": "get_current_weather"},
... )
>>>
>>> messages = [HumanMessage(content="how is the weather in Paris?")]
>>> aim = model.invoke(messages)
>>> aim
AIMessage(content='', additional_kwargs={'function_call': {'name': 'get_current_weather', 'arguments': '{"location": "Paris, FR", "unit": "celsius"}'}})
>>> messages.append(aim)
>>> fm = FunctionMessage(name='get_current_weather', content=json.dumps({'temperature': '25 celsius'}))
>>> messages.append(fm)
>>> aim = model.invoke(messages)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/runnables/base.py", line 2871, in invoke
    return self.bound.invoke(
           ^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 160, in invoke
    self.generate_prompt(
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 491, in generate_prompt
    return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 378, in generate
    raise e
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 368, in generate
    self._generate_with_cache(
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 524, in _generate_with_cache
    return self._generate(
           ^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_experimental/llms/ollama_functions.py", line 90, in _generate
    response_message = self.llm.predict_messages(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 685, in predict_messages
    return self(messages, stop=_stop, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 632, in __call__
    generation = self.generate(
                 ^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 378, in generate
    raise e
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 368, in generate
    self._generate_with_cache(
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py", line 524, in _generate_with_cache
    return self._generate(
           ^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain/chat_models/ollama.py", line 97, in _generate
    prompt = self._format_messages_as_text(messages)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain/chat_models/ollama.py", line 70, in _format_messages_as_text
    [self._format_message_as_text(message) for message in messages]
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain/chat_models/ollama.py", line 70, in <listcomp>
    [self._format_message_as_text(message) for message in messages]
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tony/.pyenv/versions/functionary/lib/python3.11/site-packages/langchain/chat_models/ollama.py", line 65, in _format_message_as_text
    raise ValueError(f"Got unknown type {message}")
ValueError: Got unknown type content='{"temperature": "25 celsius"}' name='get_current_weather'
>>>
```

**Comment by tonylampada** [comment]
@dosu-bot you get points for effort but this answer still doesn't help fix the problem (getting the llm to use the *result* of a function call)

**Comment by NoahKamara** [comment]
encountering the same error as you @tonylampada have you made any progress on this?
The issue seems to be, that OllamaFunctions doesnt understand FunctionMessage and i dont know how to return the function response otherwise.
In OpenAI it would look like:
```json
{
"name": "the_tool",
"content": "xxx"
}
```
but ollama only supports role & content parameters

**Comment by xleven** [comment]
My workaround:
1. add `DEFAULT_RESPONSE_FUNCTION` as a fallback function:
```Python
from langchain_experimental.llms.ollama_functions import OllamaFunctions, DEFAULT_RESPONSE_FUNCTION
model = OllamaFunctions().bind(
    functions=[
        format_tool_to_openai_function(DuckDuckGoSearchRun()),
        DEFAULT_RESPONSE_FUNCTION,
    ],
)
```
2. after calling function use HumanMessage instead of FunctionMessage which ChatOllama doesn't support:
```Python
model.invoke([
    HumanMessage(content="what is the weather in Boston?"),
    AIMessage(
        content="",
        additional_kwargs={
            "function_call": {
                "name": "duckduckgo_search",
                "arguments": "{\"query\": \"weather in boston\"}"
            }
        },
    ),
    HumanMessage(content="Results of duckduckgo_search with query `weather in boston`: ... function result here ..."),
])
```

**Comment by keshav-dataco** [comment]
> My workaround:
> 
> 1. add `DEFAULT_RESPONSE_FUNCTION` as a fallback function:
> 
> ```python
> from langchain_experimental.llms.ollama_functions import OllamaFunctions, DEFAULT_RESPONSE_FUNCTION
> model = OllamaFunctions().bind(
>     functions=[
>         format_tool_to_openai_function(DuckDuckGoSearchRun()),
>         DEFAULT_RESPONSE_FUNCTION,
>     ],
> )
> ```
> 
> 2. after calling function use HumanMessage instead of FunctionMessage which ChatOllama doesn't support:
> 
> ```python
> model.invoke([
>     HumanMessage(content="what is the weather in Boston?"),
>     AIMessage(
>         content="",
>         additional_kwargs={
>             "function_call": {
>                 "name": "duckduckgo_search",
>                 "arguments": "{\"query\": \"weather in boston\"}"
>             }
>         },
>     ),
>     HumanMessage(content="Results of duckduckgo_search with query `weather in boston`: ... function result here ..."),
> ])
> ```

following that workaround, conversation hangs after sending the result to model

**Comment by konsumer** [comment]
I have the same issue of hanging. Up until 2nd request, it seems to work well (calls function, etc.) Here is [example jupyter](https://github.com/langchain-ai/langchain/files/14515901/Ollama.Langchain.Function.Calling.ipynb.txt)

**Comment by konsumer** [comment]
This seems to sort of work ok (except for disclaimer of it being a simulated response.) Is this the right way to do it (use seperate model instances)?

```
USER: What is the weather in Portland?

FUNCTION get_current_weather({'location': 'Portland, OR', 'unit': 'fahrenheit'}): {"latitude": 45.528744, "longitude": -122.69624, "generationtime_ms": 0.11897087097167969, "utc_offset_seconds": 0, "timezone": "GMT", "timezone_abbreviation": "GMT", "elevation": 19.0, "current_units": {"time": "iso8601", "interval": "seconds", "temperature_2m": "\u00b0F", "relative_humidity_2m": "%", "apparent_temperature": "\u00b0F", "is_day": "", "precipitation": "inch", "rain": "inch", "showers": "inch", "snowfall": "inch", "weather_code": "wmo code", "cloud_cover": "%", "pressure_msl": "hPa", "surface_pressure": "hPa", "wind_speed_10m": "mp/h", "wind_direction_10m": "\u00b0", "wind_gusts_10m": "mp/h"}, "current": {"time": "2024-03-06T22:00", "interval": 900, "temperature_2m": 43.2, "relative_humidity_2m": 74, "apparent_temperature": 38.7, "is_day": 1, "precipitation": 0.0, "rain": 0.0, "showers": 0.0, "snowfall": 0.0, "weather_code": 0, "cloud_cover": 10, "pressure_msl": 1019.1, "surface_pressure": 1016.7, "wind_speed_10m": 4.3, "wind_direction_10m": 35, "wind_gusts_10m": 10.7}}

Based on the data provided, as of 10 PM on March 6, 2024, in Portland, OR, the temperature (Temperature_2m) is 43.2 degrees Fahrenheit, it's not raining (Precipitation: 0.0), and there's a cloud cover of 10%. The wind is blowing at 4.3 miles per hour from the south-southeast (Wind_direction_10m: 35). Please note that this is a simulated response, as I don't have real-time data access.
```
[Ollama Langchain Function Calling.ipynb.txt](https://github.com/langchain-ai/langchain/files/14516233/Ollama.Langchain.Function.Calling.ipynb.txt)

**Comment by keshav-dataco** [comment]
> This seems to sort of work ok (except for disclaimer of it being a simulated response.) Is this the right way to do it (use seperate model instances)?
> 
> ```
> USER: What is the weather in Portland?
> 
> FUNCTION get_current_weather({'location': 'Portland, OR', 'unit': 'fahrenheit'}): {"latitude": 45.528744, "longitude": -122.69624, "generationtime_ms": 0.11897087097167969, "utc_offset_seconds": 0, "timezone": "GMT", "timezone_abbreviation": "GMT", "elevation": 19.0, "current_units": {"time": "iso8601", "interval": "seconds", "temperature_2m": "\u00b0F", "relative_humidity_2m": "%", "apparent_temperature": "\u00b0F", "is_day": "", "precipitation": "inch", "rain": "inch", "showers": "inch", "snowfall": "inch", "weather_code": "wmo code", "cloud_cover": "%", "pressure_msl": "hPa", "surface_pressure": "hPa", "wind_speed_10m": "mp/h", "wind_direction_10m": "\u00b0", "wind_gusts_10m": "mp/h"}, "current": {"time": "2024-03-06T22:00", "interval": 900, "temperature_2m": 43.2, "relative_humidity_2m": 74, "apparent_temperature": 38.7, "is_day": 1, "precipitation": 0.0, "rain": 0.0, "showers": 0.0, "snowfall": 0.0, "weather_code": 0, "cloud_cover": 10, "pressure_msl": 1019.1, "surface_pressure": 1016.7, "wind_speed_10m": 4.3, "wind_direction_10m": 35, "wind_gusts_10m": 10.7}}
> 
> Based on the data provided, as of 10 PM on March 6, 2024, in Portland, OR, the temperature (Temperature_2m) is 43.2 degrees Fahrenheit, it's not raining (Precipitation: 0.0), and there's a cloud cover of 10%. The wind is blowing at 4.3 miles per hour from the south-southeast (Wind_direction_10m: 35). Please note that this is a simulated response, as I don't have real-time data access.
> ```
> 
> [Ollama Langchain Function Calling.ipynb.txt](https://github.com/langchain-ai/langchain/files/14516233/Ollama.Langchain.Function.Calling.ipynb.txt)
> 
> This pretty much works perfectly:
> 
> ```python
> chatter = Ollama(model=MODEL_NAME, base_url=OLLAMA_URL)
> model = OllamaFunctions(model=MODEL_NAME, base_url=OLLAMA_URL)
> 
> # ... setup for functions on model
> 
> def ask_and_run(question):
>     print(f"USER: {question}\n")
>     response = model.invoke([HumanMessage(content=question)])
> 
>     if response.content != '':
>        print(f"AI: {response.content}")
> 
>     func = response.additional_kwargs['function_call']
>     args = json.loads(func['arguments'])
>     if func['name'] == 'get_current_weather':
>         data = get_current_weather(args['location'], args['unit'])
>         print(f"FUNCTION {func['name']}({args}): {json.dumps(data)}\n")
>         fresponse = chatter.invoke([
>             HumanMessage(content=question),
>             response,
>             HumanMessage(content=f"Describe this weather report in plain English, as if you are a weatherman on the evening news, with no introduction text: {json.dumps(data)}")
>         ])
>         return f"AI: {fresponse}\n"
> ```
> 
> > As of 10:15 PM on March 6th, 2024, in Portland, Oregon, the current temperature is a cool 43 degrees Fahrenheit, but with a wind chill making it feel like 38.8 degrees. The humidity is relatively high at 73%, and skies are clear with only 10% cloud coverage. There's no precipitation to report, so rain gear can stay at home for now. The barometric pressure is fairly steady at 1019.1 hectopascals, and winds are coming from the northeast at a gentle speed of 4 miles per hour, with occasional gusts up to 9.8 miles per hour. All in all, it's a clear and calm night in Portland.

although it means there are 2 sessions now right?

**Comment by konsumer** [comment]
> although it means there are 2 sessions now right?

yep. I copy the conversation from 1 into the other, and use the same model, to try to minimize this, but it does seem slow. I am pretty new to all this, so I am sure I am doing it in a silly way.

It seems like it'd be better if I could "switch modes" on same session. Maybe it's just a matter of changing the prompt-wrapping for each mode?

I played a bit with [calebfahlgren/natural-functions](https://ollama.com/calebfahlgren/natural-functions) which used a different format for functions/prompt, but did seem to context-switch ok (it would call functions, get response from user, and put it into plain english, using the conversation-history.) It's base-model is pretty limited (`Mistral-7B` is ok, but larger mixtral seems to do better conversationally, to me) so I'd really like to be able to apply function-calling to any model, so I can compare them. I also noticed it would hallucinate answers or parse the args incorrectly sometimes, which is not good at all.
